doc/neps/nep-0050-scalar-promotion.rst

```diff
@@ -0,0 +1,443 @@
+   well as making it available to user defined DTypes through
+   :ref:`NEP 42 <NEP42>`.
+   Currently, this leads to a dual implementation of a new and an old (value
+   sensitive) system.  Fixing this will greatly simplify the internal logic
+   and make results more consistent.
+
+
+Usage and Impact
+================
+
+There will be no transition period due to the difficulty and noise this is
+expected to create.  In rare cases users may need to adjust code to avoid
+reduced precision or incorrect results.
+
+We plan to provide an *optional* warning mode capable of notifying users of
+potential changes in behavior in most relevant cases.
+
+
+Impact on ``can_cast``
+----------------------
+
+Can cast will never inspect the value anymore.  So that the following results
+are expected to change from ``True`` to ``False``::
+
+  np.can_cast(100, np.uint8)
+  np.can_cast(np.int64(100), np.uint8)
+  np.can_cast(np.array(100, dtype=np.int64), np.uint8)
+
+We expect that the impact of this change will be small compared to that of
+the following changes.
+
+.. note::
+
+    The first example where the input is a Python scalar could be preserved
+    to some degree, but this is not currently planned.
+
+
+Impact on operators functions involving NumPy arrays or scalars
+---------------------------------------------------------------
+
+The main impact on operations not involving Python scalars (float, int, complex)
+will be that 0-D arrays and NumPy scalars will never behave value-sensitive.
+This removes currently surprising cases.  For example::
+
+  np.arange(10, dtype=np.uint8) + np.int64(1)
+
+Will return an int64 array because the type of ``np.int64(1)`` is strictly
+honoured.
+
+
+Impact on operators involving Python ``int``, ``float``, and ``complex``
+------------------------------------------------------------------------
+
+This NEP attempts to preserve most of the convenience that the old behaviour
+gave for Python operators, but remove it for
+
+The current value-based logic has some nice properties when "untyped" Python
+scalars involved::
+
+  np.arange(10, dtype=np.int8) + 1  # returns an int8 array
+  nparray([1., 2.], dtype=np.float32) * 3.5  # returns a float32 array
+
+But led to complexity when it came to "unrepresentable" values:
+
+  np.arange(10, dtype=np.int8) + 256  # returns int16
+  nparray([1., 2.], dtype=np.float32) * 1e200  # returns float64
+
+The proposal is to preserve this behaviour for the most part.  This is achieved
+by considering Python ``int``, ``float``, and ``complex`` to be "weakly" typed
+in these operations.
+To mitigate user surprises, we would further make conversion to the new type
+more strict.  This means that the results will be unchanged in the first
+two examples.  For the second one, the results will be the following::
+
+  np.arange(10, dtype=np.int8) + 256  # raises a TypeError
+  nparray([1., 2.], dtype=np.float32) * 1e200  # warning and returns infinity
```

Comment:
`np.array`

Reply:
Not sure how much value there is in it, but for the second changed case that returns infinity, are there examples from other frameworks/programming languages that do something similar in a matching scenario?

I don't know that we'd necessarily want to mimic what other major languages/projects do, though citing them as examples may help make a stronger case if needed. Perhaps in this case one could just quickly verify in C or C++ or whatever.

Reply:
I added an `admonition`, C normally does this, but of course you can set the evironment to raise an error (just like you can in NumPy, although casts currently don't check that normally!).

I decided to promise to fix that it gives a warning right now.  `pytorch` seems to just overflow silently, I guess they probably don't give warnings reliably, because some hardware they care about doesn't really support it well.